### PR TITLE
Fixed #31370 -- Added support for parallel tests with --buffer.

### PR DESCRIPTION
**Pull request by adamchainz**
[ticket](https://code.djangoproject.com/ticket/31370#ticket)

### CODE REVIEW

**Comment by adamchainz** [comment]
I've also added a commit to make the `runtests.py` script always use buffer mode.

Tbh I'm in favour of making buffer the default for all users. We could add an `--unbuffer` option to flip it off. This is the default behaviour of pytest and it can speed up tests quite a bit since writing output to screen can become a limiting factor.

**Comment by adamchainz** [comment]
Paging @bmispelon for review since you wrote the first buffer PR #12169.

File: django/test/runner.py [hunk 1]
```diff
@@ -114,9 +116,24 @@ def __init__(self):
 
         self.events = []
         self.failfast = False
+        self.buffer = False
         self.shouldStop = False
+        self._stdout_buffer = None
+        self._stderr_buffer = None
+        self._original_stdout = sys.stdout
+        self._original_stderr = sys.stderr
+        self._mirrorOutput = False
```
Comment (adamchainz) [inline django/test/runner.py]: Most of the changes to this class are copied from unittest's `TextResult`: https://github.com/python/cpython/blob/a45b695b9fcfbbb0a087222abc5c8d691a7d2770/Lib/unittest/result.py#L47-L52 . But it was already half copied.

Comment (bmispelon) [reply django/test/runner.py]: > Paging @bmispelon for review since you wrote the first buffer PR #12169.

I don't feel very confident reviewing this. My work on #12169 was mostly to write glue code to pass the option correctly to python's unittest.

One thing I wonder is whether you've investigated actually extending the base `TestResult` instead of copy/pasting from it.
The docstring of `RemoteTestResult` seems to indicate that the copy/pasting was done because Django didn't need the whole API of `TestResult` but I get the feeling that we're trying to implement that API now, so maybe it'd be worth it to extend rather than copy/paste.

Relying on Python's implementation would also make me more confident about the lack of tests.

Comment (adamchainz) [reply django/test/runner.py]: > One thing I wonder is whether you've investigated actually extending the base TestResult instead of copy/pasting from it.

I had not. I guess I preferred to leave the sleeping dog lie, but after further inspection it looks pretty easy. It in fact removes a lot of duplication on our side such as the stop flag and method.

I've "neutered" a method and the list data structures to avoid unnecessarily increasing processing time and memory usage, see comments.

File: django/test/runner.py [hunk 2]
```diff
@@ -97,25 +97,47 @@ def debug(self, error):
...
-    inherit unittest.TestResult and doesn't attempt to implement all its API.
+    def append(self, item):
+        pass
 
-    The implementation matches the unpythonic coding style of unittest2.
+
+class RemoteTestResult(unittest.TestResult):
+    """
+    Extend unittest.TestResult to record events in the child processes so they
+    can be replayed in the master process. Events include things like which
```
Comment (bmispelon) [inline django/test/runner.py]: How about using "parent" here instead of "master"? Seems like a better fit with "child processes".

File: django/test/runner.py [hunk 3]
```diff
@@ -255,10 +276,22 @@ def addExpectedFailure(self, test, err):
...
+        super().addExpectedFailure(test, err)
 
     def addUnexpectedSuccess(self, test):
         self.events.append(('addUnexpectedSuccess', self.test_index))
-        self.stop_if_failfast()
+        super().addUnexpectedSuccess(test)
+
+    def wasSuccessful(self):
+        """Tells whether or not this result was a success."""
+        failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)
```
Comment (bmispelon) [inline django/test/runner.py]: A set literal makes a tiny bit more sense here I think: `{'addError', 'addFailure', 'addUnexpectedSuccess',}`

File: django/test/runner.py (hunk 2 above)
Comment (adamchainz) [reply django/test/runner.py]: Yes and I remember the master/slave controversy.

File: django/test/runner.py (hunk 3 above)
Comment (adamchainz) [reply django/test/runner.py]: Yes it does make sense and I've changed it.

Fun fact though is that a tuple is a tiny bit faster because it can be cached in the function definition, whilst the set needs to be reconstructed each time it runs:

```python
In [2]: x = 'addUnexpectedSuccess'

In [3]: %timeit x in failure_types
54.6 ns ± 0.318 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: failure_types = {'addError', 'addFailure', 'addUnexpectedSuccess'}

In [5]: %timeit x in failure_types
35.6 ns ± 0.415 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: def test_tuple(x):
   ...:     failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)
   ...:     return x in failure_types
   ...:

In [7]: def test_set(x):failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)

In [8]: def test_set(x):failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)

In [9]: def test_set(x):failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)

In [10]: def test_set(x):
    ...:     failure_types = {'addError', 'addFailure', 'addUnexpectedSuccess'}
    ...:     return x in failure_types
    ...:
```

In [11]: %timeit test_tuple(x)
120 ns ± 0.585 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [12]: %timeit test_set(x)
161 ns ± 0.677 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

Comment (bmispelon) [reply django/test/runner.py]: Oh, interesting. Is that because of mutability? What about a frozenset then? I'm not suggesting to change it here (I don't think it matters much) but you made me curious now :)

Comment (adamchainz) [reply django/test/runner.py]: It's an immutability optimization yes

Unfortunately frozensets don't work since they aren't a literal type. You can see here only the tuple comes through `LOAD_CONST` - one of the "constants" for the function:

```python
In [5]: dis.dis(test_tuple)
  2           0 LOAD_CONST               1 (('addError', 'addFailure', 'addUnexpectedSuccess'))
              2 STORE_FAST               1 (failure_types)

  3           4 LOAD_FAST                0 (x)
              6 LOAD_FAST                1 (failure_types)
              8 COMPARE_OP               6 (in)
             10 RETURN_VALUE

In [6]: dis.dis(test_set)
  2           0 LOAD_CONST               1 ('addError')
              2 LOAD_CONST               2 ('addFailure')
              4 LOAD_CONST               3 ('addUnexpectedSuccess')
              6 BUILD_SET                3
              8 STORE_FAST               1 (failure_types)

  3          10 LOAD_FAST                0 (x)
             12 LOAD_FAST                1 (failure_types)
             14 COMPARE_OP               6 (in)
             16 RETURN_VALUE

In [7]: def test_frozenset(x):
   ...:     failure_types = frozenset({'addError', 'addFailure', 'addUnexpectedSuccess'})
   ...:     return x in failure_types
   ...:

In [8]: dis.dis(test_frozenset)
  2           0 LOAD_GLOBAL              0 (frozenset)
              2 LOAD_CONST               1 ('addError')
              4 LOAD_CONST               2 ('addFailure')
              6 LOAD_CONST               3 ('addUnexpectedSuccess')
              8 BUILD_SET                3
             10 CALL_FUNCTION            1
             12 STORE_FAST               1 (failure_types)

  3          14 LOAD_FAST                0 (x)
             16 LOAD_FAST                1 (failure_types)
             18 COMPARE_OP               6 (in)
             20 RETURN_VALUE

```

**Comment by adamchainz** [comment]
Rebased. I also made a ticket to propose making buffering the default, with a `--no-buffer` option to disable it: https://code.djangoproject.com/ticket/31543#ticket .

**Comment by adamchainz** [comment]
@Valze you may also be interested to see or review this PR, since it's test runner related.

File: django/test/runner.py [hunk 4]
```diff
@@ -97,25 +97,51 @@ def debug(self, error):
...
+    Dummy list class for faking storage of results in unittest.TestResult. We
+    do this because RemoteTestResult replaces the way of recording.
+    """
+    __slots__ = []
 
-    The sole purpose of this class is to record events in the child processes
-    so they can be replayed in the master process. As a consequence it doesn't
-    inherit unittest.TestResult and doesn't attempt to implement all its API.
+    def append(self, item):
+        pass
```
Comment (adamchainz) [inline django/test/runner.py:108]: not sure if it's better to just replace with a `mock.Mock()`.

**Comment by Valze** [comment]
> @Valze you may also be interested to see or review this PR, since it's test runner related.

I've been keeping an eye on it ever since you started solving it initially a month ago. I'll give it a more thorough read over this weekend and see if I can help in any way.

File: django/test/runner.py (hunk 4 above)
Comment (Valze) [reply django/test/runner.py:108]: I don't think it would make a difference in memory efficiency. Seems fine to keep it as it is.

**Comment by Valze** [comment]
Code looks much cleaner after subclassing TextResult, and it really does improve a lot of other written methods.

I'm +1 on setting buffer as the default option. It removes unnecessary clutter

File: django/test/runner.py (hunk 4 above)
Comment (adamchainz) [reply django/test/runner.py:108]: On second thoughts, it would be less efficient that a list. Mocks record a list of all calls made against them.

I also admit that replacing the lists is even a *bit* of a premature optimization. In large projects I imagine there could be thousands of test results stored in memory, which could translate at best to a few megabytes of memory usage.

**Comment by adamchainz** [comment]
> I'm +1 on setting buffer as the default option. It removes unnecessary clutter

Thanks. Can you repeat that on [the ticket](https://code.djangoproject.com/ticket/31543#ticket) ?

File: django/test/runner.py [hunk 5]
```diff
@@ -97,25 +97,51 @@ def debug(self, error):
...
 
-    The implementation matches the unpythonic coding style of unittest2.
+
+class RemoteTestResult(unittest.TestResult):
```
Comment (felixxm) [inline django/test/runner.py:111]: Can we move subclassing `unittest.TestResult` to a separate commit? Is it a valuable optimization on its own? or it's strictly related with supporting `--buffer` with `--parallel`.

File: django/test/runner.py [hunk 6]
```diff
@@ -255,10 +280,22 @@ def addExpectedFailure(self, test, err):
...
+        super().addUnexpectedSuccess(test)
+
+    def wasSuccessful(self):
+        """Tells whether or not this result was a success."""
+        failure_types = {'addError', 'addFailure', 'addUnexpectedSuccess'}
+        return all(e[0] not in failure_types for e in self.events)
+
+    def _exc_info_to_string(self, err, test):
+        # Deliberately neuter this method. It only powers the default unittest
+        # behavior for recording errors - we don't use it here since we instead
```
Comment (felixxm) [inline django/test/runner.py:296]: Try to avoid `we` in comments.

File: tests/test_runner/test_discover_runner.py [hunk 7]
```diff
@@ -269,14 +269,6 @@ def test_pdb_with_parallel(self):
...
             DiscoverRunner(pdb=True, parallel=2)
 
-    def test_buffer_with_parallel(self):
-        msg = (
-            'You cannot use -b/--buffer with parallel tests; pass '
-            '--parallel=1 to use it.'
-        )
-        with self.assertRaisesMessage(ValueError, msg):
-            DiscoverRunner(buffer=True, parallel=2)
-
```
Comment (felixxm) [inline tests/test_runner/test_discover_runner.py:279]: Can we add a test? similar to the `test_buffer_mode_test_pass()`.

File: tests/test_runner/test_parallel.py [hunk 8]
```diff
@@ -52,6 +53,39 @@ def dummy_test(self):
...
+
+    def test_was_successful_one_failure(self):
+        result = RemoteTestResult()
+        result.addFailure(None, ValueError("woops"))
+        self.assertFalse(result.wasSuccessful())
+
+    def test_picklable(self):
+        result = RemoteTestResult()
+        pickle.loads(pickle.dumps(result))
+
```
Comment (felixxm) [inline tests/test_runner/test_parallel.py:88]: This works without a patch, so we can move it to a separate commit.

**Comment by felixxm** [comment]
@adamchainz Do you have time to check my comments and polish this patch?

File: tests/test_runner/test_discover_runner.py (hunk 7 above)
Comment (adamchainz) [reply tests/test_runner/test_discover_runner.py:279]: I tried:

```python
    def test_buffer_mode_parallel(self):
        runner = DiscoverRunner(parallel=2, buffer=True, verbose=0)
        with captured_stdout() as stdout, captured_stderr() as stderr:
            suite = runner.build_suite([
                'test_runner_apps.buffer.tests_buffer.WriteToStdoutStderrTestCase.test_pass',
                'test_runner_apps.buffer.tests_buffer.WriteToStdoutStderrTestCase2.test_pass',
            ])
            runner.run_suite(suite)
        self.assertNotIn('Write to stderr.', stderr.getvalue())
        self.assertNotIn('Write to stdout.', stdout.getvalue())
```

But this doesn't work because we can't capture stdin/stdout across the parallel processes. Any suggestions?

File: tests/test_runner/test_parallel.py (hunk 8 above)
Comment (adamchainz) [reply tests/test_runner/test_parallel.py:88]: Done

File: django/test/runner.py (hunk 5 above)
Comment (adamchainz) [reply django/test/runner.py:111]: Done

**Comment by adamchainz** [comment]
I finally got back to this.

File: tests/test_runner/test_parallel.py (hunk 8 above)
Comment (adamchainz) [reply tests/test_runner/test_parallel.py:87]: I think you meant just the picklable test here, since the other tests cover functionality added later when changing the inheritance of `RemoteTestResult`, right?

Comment (felixxm) [reply tests/test_runner/test_parallel.py:87]: Yes.

**Comment by felixxm** [comment]
@adamchainz Thanks for updates :+1: I still believe that we should drop the last commit and keep the current (default) behavior for `runtests.py`, see [comment](https://github.com/django/django/pull/12580#pullrequestreview-458161677).

@carltongibson Can I ask for your opinion?

**Comment by adamchainz** [comment]
I'm happy to drop the last commit if you think so, but I have not seen any pytest users complain about its automatic buffering. You can always do `print()` + `assert 0` to see the output.

**Review by carltongibson — COMMENTED** [review]
Yes, please, can we not change the behaviour of `./runtests.py` here. (`-s`  is amongst my favourite pytest options.)

Otherwise looks good.

**Comment by adamchainz** [comment]
Okay I dropped the change to `runtests.py`

File: django/test/runner.py [hunk 9]
```diff
@@ -264,10 +286,22 @@ def addExpectedFailure(self, test, err):
...
+        super().addExpectedFailure(test, err)
 
     def addUnexpectedSuccess(self, test):
         self.events.append(('addUnexpectedSuccess', self.test_index))
-        self.stop_if_failfast()
+        super().addUnexpectedSuccess(test)
+
+    def wasSuccessful(self):
+        """Tells whether or not this result was a success."""
+        failure_types = {'addError', 'addFailure', 'addUnexpectedSuccess'}
```
Comment (felixxm) [inline django/test/runner.py:297]: I think we should add `addSubTest` to this set. I added an assertion that fails.

**Review by felixxm — COMMENTED** [review]
@adamchainz Thanks for updates :+1: I pushed edits to the first two commits and noticed one issue.

**Comment by felixxm** [comment]
@adamchainz Also, it doesn't really work for me. I think you missed passing `buffer` to the `RemoteTestResult`. I pushed commit with a proposed fix.

**Comment by adamchainz** [comment]
>  I think you missed passing buffer to the RemoteTestResult. I pushed commit with a proposed fix.

Ah thanks. I think I accidentally dropped that in the rebase-and-split.

File: django/test/runner.py (hunk 9 above)
Comment (adamchainz) [reply django/test/runner.py]: Yup, well spotted!

**Comment by felixxm** [comment]
I noticed that signature are outdated, see #14145.

**Comment by felixxm** [comment]
@adamchainz Thanks :+1: It's ready to go :star2:

**Comment by adamchainz** [comment]
🙌 🚀
